### ISSUE TITLE
Custom loader without using Request and Decoder

### ISSUE THREAD

**Issue by ealymbaev** [kean/Nuke#133]
Hi,

First of all, thank you for the great library, so far it works like a charm.

I wanted to use it for my Avatar Views. First let me explain the logic: avatar view should show the image from user's picture url, BUT if there is no url - it should show placeholder image. Please note that the image is generated depending on the user's id. (different colors).

The best option for me would be the following:

- Some custom Loader that generates images (based on id) and applies Processors.
- This loader should be able to work without Request's, as I don't need any URL's. All I need is to provide some id for generating correct image color.
- Thus, all the images (for different colors) would be generated only once and stored in cache. After that they could be reused from cache applying different processor (resize, rounded corners).

In short, I need an option to use generated images instead of images from url request.

**Comment by kean** [comment]
Hey! Sorry, really busy at work right now. My initial thoughts are that Nuke is not designed to support this. It's better If you just implement this aside without using Nuke infrastructure.

**Comment by ealymbaev** [comment]
Strictly binding Nuke to Requests makes it quite difficult to extend. And Dependency Injection principle does not work here, as I cannot substitute any layer with my own, without binding to Request class.

It would be great for library if every layer could work without depending on other one. I  other words as a developer I want to be able to use Caching and Processing layers without the need of any data loading layers.

But you have integrated processors into Request class and also require Request class to be present in Caching and Processing layer, but they don't need any URL actually, the just transform and store images in cache.

**Comment by ealymbaev** [comment]
Initially I wanted to take your library as a start point and implement all my ideas above myself, but I don't want to loose your further updates.

**Comment by kean** [comment]
Strictly speaking you can:
- use `Processing` protocol and any type that implements it without `Request`
- use `Cache` without `Request`, but you'll need to provide your own cache keys

I get what you mean tho, yes there is some room for improvement in terms of decoupling components. I might invest in this area some time in the future. If you have some specific ideas please let me know, I might implement them for Nuke 6.

I'm personally a bit reluctant to change anything at this point. The core framework works great, it's simple, stable, performant and does just what I need in my apps. If I need something specific (for example, the case you've described in this issue) I would just grab RxSwift and implement it. To be honest, RxSwift makes about 50% of Nuke code/features obsolete.
